### PR TITLE
ARROW-6622: [R] Normalize paths for filesystem API on Windows

### PR DESCRIPTION
**Pull request by nealrichardson**
I fixed this in R, though I wonder if that's not enough. At a minimum, the C++ docs should note this requirement (@pitrou maybe you know the best place in the docs to add this?), and I still think it would be nice if all of this path normalization were handled in C++ (cf. https://issues.apache.org/jira/browse/ARROW-6324).

### CODE REVIEW

**Comment by wesm** [comment]
I think I'm with @pitrou here that introducing automatic path normalization into the C++ filesystem API might be a bit too much scope creep -- we could provide a built-in path normalization function for consumers that don't have it (e.g. Python has pathlib, R has its own function here)

**Comment by nealrichardson** [comment]
I can accept that, though we should document the C++ API's requirements accordingly. I created https://issues.apache.org/jira/browse/ARROW-6629 because I didn't see any filesystem docs at all, so nowhere to just add a note in this patch.

**Comment by nealrichardson** [comment]
This is failing, though differently than before: https://ci.appveyor.com/project/nealrichardson/arrow/builds/27536178/job/gu10af0vck7gum81#L4342

The path normalization is making absolute paths with respect to the current working directory in the R process, and that can clash with whatever the FileSystem object has as its path. I'll need to rethink the approach.

**Comment by nealrichardson** [comment]
This is passing and ready for review @romainfrancois

File: r/R/filesystem.R [hunk 1]
```diff
@@ -232,6 +243,17 @@ LocalFileSystem$create <- function() {
 #' @export
 SubTreeFileSystem <- R6Class("SubTreeFileSystem", inherit = FileSystem)
 SubTreeFileSystem$create <- function(base_path, base_fs) {
-  xp <- fs___SubTreeFileSystem__create(base_path, base_fs)
+  xp <- fs___SubTreeFileSystem__create(fs_relative_path(base_path), base_fs)
   shared_ptr(SubTreeFileSystem, xp)
 }
+
+clean_path <- function(path) {
```
Comment (romainfrancois) [inline r/R/filesystem.R]: This looks like things potentially better fixed with `fs::` cc @jimhester. I know `fs` has been downgraded to `Suggests` here, but I'm just worried we'd be missing cases with the implementation below.

**Review by romainfrancois — COMMENTED** [review]
Should we use `fs::` here ?

File: r/R/filesystem.R [hunk 2]
```diff
@@ -232,6 +243,17 @@ LocalFileSystem$create <- function() {
...
+  # Make sure we have a valid, absolute, forward-slashed path for passing to Arrow
+  normalizePath(path, winslash = "/", mustWork = FALSE)
+}
+
+fs_relative_path <- function(path) {
+  # Make sure all path separators are "/", not "\" as on Windows
+  path_sep <- ifelse(tolower(Sys.info()[["sysname"]]) == "windows", "\\\\", "/")
```
Comment (jimhester) [inline r/R/filesystem.R]: You should probably use `.Platform$file.sep`

Comment (jimhester) [reply r/R/filesystem.R]: Also not sure about the name of this function, it is not really doing anything to compute a relative path, you could probably just use `normalizePath(mustWork = FALSE)`, assuming path doesn't actually have to exist on the filesystem.

File: r/R/filesystem.R [hunk 3]
```diff
@@ -108,7 +108,10 @@ Selector <- R6Class("Selector",
 )
 
 Selector$create <- function(base_dir, allow_non_existent = FALSE, recursive = FALSE) {
-  shared_ptr(Selector, fs___Selector__create(base_dir, allow_non_existent, recursive))
+  shared_ptr(
+    Selector,
+    fs___Selector__create(fs_relative_path(base_dir), allow_non_existent, recursive)
```
Comment (pitrou) [inline r/R/filesystem.R]: Are you sure? Will it work correctly if `base_dir` is e.g. a path in an S3 filesystem, or will it produce bogus results?

Comment (pitrou) [reply r/R/filesystem.R]: Same question for all such cases below.

**Review by pitrou — COMMENTED** [review]
This doesn't look correct to me. The simple solution should be to start with /-separated slashes and only use that.

File: r/R/filesystem.R (hunk 1 above)
Comment (nealrichardson) [reply r/R/filesystem.R]: I didn't see the `fs` function that does the winslash normalization we need here, but more than happy to use it if it exists.

File: r/tests/testthat/test-filesystem.R [hunk 4]
```diff
@@ -83,7 +83,6 @@ test_that("SubTreeFilesystem", {
   expect_is(st_fs, "FileSystem")
```
Comment (pitrou) [inline r/tests/testthat/test-filesystem.R]: So you should simply /-normalize `td` above. Should be easy :-)

File: r/R/filesystem.R (hunk 2 above)
Comment (nealrichardson) [reply r/R/filesystem.R]: Re: `.Platform$file.sep`, I intended to but then on my Windows VM, `.Platform$file.sep` reported "/" while `tempfile()` still created paths with "\"  🤷‍♂  

I originally did `normalizePath` but unfortunately for this use case, `normalizePath` does path expansion, even with `mustWork=FALSE`. We need to keep the path relative (to whatever the FileSystem object holds as its cwd) and just need Windows `\` to be converted to `/`. Regular expressions were my last resort and I'm happy to use something better or more robust if it exists.

**Comment by pitrou** [comment]
Also note that `SubTreeFileSystem` is not really needed by end-users. It's mostly useful for internal C++ testing, because there is a sizable generic filesystem test suite that expects to write at the root of an empty filesystem.

File: r/tests/testthat/test-filesystem.R (hunk 4 above)
Comment (nealrichardson) [reply r/tests/testthat/test-filesystem.R]: Should be!

Comment (pitrou) [reply r/tests/testthat/test-filesystem.R]: At worse you simply string-replace `\` with `/` (but only when on Windows :-)).

Comment (nealrichardson) [reply r/tests/testthat/test-filesystem.R]: That's where I ended up: https://github.com/apache/arrow/pull/5445/files#diff-3bd936ee39b8abe149e5399681610c75R255

Comment (pitrou) [reply r/tests/testthat/test-filesystem.R]: Ah, then you shouldn't call it `fs_relative_path`...

File: r/R/filesystem.R (hunk 2 above)
Comment (pitrou) [reply r/R/filesystem.R]: It looks a bit weird to do a no-op replacement on Unix, though I suppose it doesn't work.

**Comment by nealrichardson** [comment]
I renamed the functions to `clean_path_abs()` and `clean_path_rel()`. Hopefully that makes their intent and output more clear. 

I wouldn't be surprised if we had to revisit this in the future, but for now this seems like a step forward, so I'll merge unless there are objections.

**Pull request closed without merging**